src/java.base/share/classes/java/io/InputStream.java

```diff
@@ -542,7 +547,12 @@ public long skip(long n) throws IOException {
         }
 
         int size = (int)Math.min(MAX_SKIP_BUFFER_SIZE, remaining);
-        byte[] skipBuffer = new byte[size];
+
+        byte[] skipBuffer = this.skipBuffer;
+        if ((skipBuffer == null) || (skipBuffer.length < size)) {
+            this.skipBuffer = skipBuffer = new byte[size < MIN_SKIP_BUFFER_SIZE ? MIN_SKIP_BUFFER_SIZE : MAX_SKIP_BUFFER_SIZE];
+        }
+
         while (remaining > 0) {
             nr = read(skipBuffer, 0, (int)Math.min(size, remaining));
```

Comment:
The check for `skipBuffer.length < size` makes it appear that the buffer can be re-allocated.
If it is allocated once then only the `skipBuffer == null` is needed.

The code may be simpler if the 'size' variable is removed.
```
        byte[] skipBuffer = this.skipBuffer;
        if (skipBuffer == null) {
            this.skipBuffer = skipBuffer = 
                new byte[(remaining < MIN_SKIP_BUFFER_SIZE) ? MIN_SKIP_BUFFER_SIZE : MAX_SKIP_BUFFER_SIZE];
        }
        while (remaining > 0) {
            int nr = read(skipBuffer, 0, (int)Math.min(skipBuffer.length, remaining));
```

Reply:
It indeed is reallocated when the existing one is not large enough.

Reply:
> Sorry, I misunderstood your earlier comment: "Sounds reasonable and applied" as concurring with not re-allocating to avoid the overhead of gc and wasting the smaller buffer. The code can be harder to understand because `size` and `skipBuffer.length` may be different.

Rethink of this, I kind of prefer the soft reference plan @liach shown....